File: contracts/identifiers/python/canonical_identifiers.py
"""Canonical StackSniffer identifier bindings."""

from __future__ import annotations

from dataclasses import dataclass, fields
import hashlib
import json
import re
from typing import Any, ClassVar
from uuid import UUID, uuid4

_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
_REPOSITORY_KEY_RE = re.compile(r"^[a-z][a-z0-9-]{1,31}/[a-z0-9][a-z0-9_.-]{0,99}/[a-z0-9][a-z0-9_.-]{0,99}$")
_TAXONOMY_VERSION_RE = re.compile(r"^v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$")

_ID_FIELDS = (
    "repository_version_id", "analysis_id", "analysis_job_id",
    "classification_run_id", "classification_result_id", "agent_run_id",
    "event_id", "request_id", "correlation_id",
)
_COMMIT_SHA_RE = re.compile(r"^[0-9a-f]{7,64}$")
_HASH_PREFIX = "sha256:"


def _validate_uuid(field_name: str, value: str) -> str:
    if not isinstance(value, str) or not _UUID_RE.fullmatch(value):
        raise ValueError(f"{field_name} must be a canonical lowercase UUID")
    try:
        UUID(value)
    except ValueError as exc:
        raise ValueError(f"{field_name} must be a valid UUID") from exc
    return value


def validate_repository_key(value: str) -> str:
    if not isinstance(value, str) or not _REPOSITORY_KEY_RE.fullmatch(value):
        raise ValueError("repository_key must be provider/owner/repository in lowercase")
    return value


def validate_taxonomy_version(value: str) -> str:
    if not isinstance(value, str) or not _TAXONOMY_VERSION_RE.fullmatch(value):
        raise ValueError("taxonomy_version must use vMAJOR.MINOR.PATCH")
    return value


def _validate_text(field_name: str, value: str) -> str:
    if not isinstance(value, str) or not value:
        raise ValueError(f"{field_name} must be a non-empty string")
    return value


def _validate_commit_sha(value: str) -> str:
    if not isinstance(value, str) or not _COMMIT_SHA_RE.fullmatch(value):
        raise ValueError("commit_sha must be 7-64 lowercase hexadecimal characters")
    return value


def _derive(namespace: str, values: dict[str, str]) -> str:
    payload = {"namespace": namespace, **values}
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
    digest = hashlib.sha256(canonical.encode("utf-8")).hexdigest()
    return f"{_HASH_PREFIX}{digest}"


def deterministic_analysis_key(repository_key: str, commit_sha: str,
                               deterministic_pipeline_version: str) -> str:
    validate_repository_key(repository_key)
    _validate_commit_sha(commit_sha)
    _validate_text("deterministic_pipeline_version", deterministic_pipeline_version)
    return _derive("deterministic_analysis", {
        "commit_sha": commit_sha,
        "deterministic_pipeline_version": deterministic_pipeline_version,
        "repository_key": repository_key,
    })


def classification_contract_id(classification_pipeline_version: str,
                               classification_model_profile: str,
                               prompt_version: str,
                               taxonomy_version: str) -> str:
    _validate_text("classification_pipeline_version", classification_pipeline_version)
    _validate_text("classification_model_profile", classification_model_profile)
    _validate_text("prompt_version", prompt_version)
    validate_taxonomy_version(taxonomy_version)
    return _derive("classification_contract", {
        "classification_model_profile": classification_model_profile,
        "classification_pipeline_version": classification_pipeline_version,
        "prompt_version": prompt_version,
        "taxonomy_version": taxonomy_version,
    })


def classification_key(analysis_id: str, contract_id: str) -> str:
    _validate_uuid("analysis_id", analysis_id)
    _validate_text("classification_contract_id", contract_id)
    return _derive("classification", {
        "analysis_id": analysis_id,
        "classification_contract_id": contract_id,
    })


def assembled_analysis_key(deterministic_key: str, classification_key_value: str) -> str:
    _validate_text("deterministic_analysis_key", deterministic_key)
    _validate_text("classification_key", classification_key_value)
    return _derive("assembled_analysis", {
        "classification_key": classification_key_value,
        "deterministic_analysis_key": deterministic_key,
    })


def agent_review_spec_key(repository_key: str, commit_sha: str, review_goal: str,
                          agent_version: str, prompt_version: str,
                          model_profile: str) -> str:
    validate_repository_key(repository_key)
    _validate_commit_sha(commit_sha)
    for field_name, value in (("review_goal", review_goal), ("agent_version", agent_version),
                              ("prompt_version", prompt_version), ("model_profile", model_profile)):
        _validate_text(field_name, value)
    return _derive("agent_review_spec", {
        "agent_version": agent_version,
        "commit_sha": commit_sha,
        "model_profile": model_profile,
        "prompt_version": prompt_version,
        "repository_key": repository_key,
        "review_goal": review_goal,
    })


def agent_context_fingerprint(classification_result_id: str, retrieval_index_version: str,
                              taxonomy_version: str, graph_snapshot_version: str = "NONE") -> str:
    _validate_uuid("classification_result_id", classification_result_id)
    _validate_text("retrieval_index_version", retrieval_index_version)
    validate_taxonomy_version(taxonomy_version)
    _validate_text("graph_snapshot_version", graph_snapshot_version)
    return _derive("agent_context", {
        "classification_result_id": classification_result_id,
        "graph_snapshot_version": graph_snapshot_version,
        "retrieval_index_version": retrieval_index_version,
        "taxonomy_version": taxonomy_version,
    })


def agent_result_cache_key(review_spec_key: str, context_fingerprint: str) -> str:
    _validate_text("agent_review_spec_key", review_spec_key)
    _validate_text("agent_context_fingerprint", context_fingerprint)
    return _derive("agent_result_cache", {
        "agent_context_fingerprint": context_fingerprint,
        "agent_review_spec_key": review_spec_key,
    })


@dataclass(frozen=True)
class IdentifierBundle:
    repository_key: str
    repository_version_id: str
    analysis_id: str
    analysis_job_id: str
    classification_run_id: str
    classification_result_id: str
    agent_run_id: str
    event_id: str
    request_id: str
    correlation_id: str
    taxonomy_version: str

    def __post_init__(self) -> None:
        validate_repository_key(self.repository_key)
        for field_name in _ID_FIELDS:
            _validate_uuid(field_name, getattr(self, field_name))
        validate_taxonomy_version(self.taxonomy_version)

    @classmethod
    def new(cls, repository_key: str, taxonomy_version: str) -> "IdentifierBundle":
        values = {field_name: str(uuid4()) for field_name in _ID_FIELDS}
        return cls(repository_key=repository_key, taxonomy_version=taxonomy_version, **values)

    def to_dict(self) -> dict[str, str]:
        return {field.name: getattr(self, field.name) for field in fields(self)}

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), sort_keys=True, separators=(",", ":"))

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "IdentifierBundle":
        expected = {field.name for field in fields(cls) if field.name != "_FIELD_NAMES"}
        if not isinstance(payload, dict) or set(payload) != expected:
            raise ValueError("identifier payload must contain exactly the canonical fields")
        return cls(**payload)

    @classmethod
    def from_json(cls, payload: str) -> "IdentifierBundle":
        try:
            decoded = json.loads(payload)
        except json.JSONDecodeError as exc:
            raise ValueError("identifier payload must be valid JSON") from exc
        return cls.from_dict(decoded)

File: contracts/identifiers/java/CanonicalIdentifiers.java
package io.stacksniffer.contracts.identifiers;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.UUID;
import java.util.regex.Pattern;

/** Canonical StackSniffer identifier bindings and JSON-field serialization. */
public final class CanonicalIdentifiers {
    private static final Pattern UUID_PATTERN = Pattern.compile(
            "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$");
    private static final Pattern REPOSITORY_KEY_PATTERN = Pattern.compile(
            "^[a-z][a-z0-9-]{1,31}/[a-z0-9][a-z0-9_.-]{0,99}/[a-z0-9][a-z0-9_.-]{0,99}$");
    private static final Pattern TAXONOMY_VERSION_PATTERN = Pattern.compile(
            "^v(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)$");

    private CanonicalIdentifiers() {
    }

    public record Bundle(
            String repository_key,
            String repository_version_id,
            String analysis_id,
            String analysis_job_id,
            String classification_run_id,
            String classification_result_id,
            String agent_run_id,
            String event_id,
            String request_id,
            String correlation_id,
            String taxonomy_version) {
        public Bundle {
            validateRepositoryKey(repository_key);
            validateUuid("repository_version_id", repository_version_id);
            validateUuid("analysis_id", analysis_id);
            validateUuid("analysis_job_id", analysis_job_id);
            validateUuid("classification_run_id", classification_run_id);
            validateUuid("classification_result_id", classification_result_id);
            validateUuid("agent_run_id", agent_run_id);
            validateUuid("event_id", event_id);
            validateUuid("request_id", request_id);
            validateUuid("correlation_id", correlation_id);
            validateTaxonomyVersion(taxonomy_version);
        }

        public static Bundle create(String repositoryKey, String taxonomyVersion) {
            return new Bundle(repositoryKey, uuid(), uuid(), uuid(), uuid(), uuid(), uuid(), uuid(), uuid(), uuid(), taxonomyVersion);
        }

        public Map<String, String> toMap() {
            Map<String, String> values = new LinkedHashMap<>();
            values.put("repository_key", repository_key);
            values.put("repository_version_id", repository_version_id);
            values.put("analysis_id", analysis_id);
            values.put("analysis_job_id", analysis_job_id);
            values.put("classification_run_id", classification_run_id);
            values.put("classification_result_id", classification_result_id);
            values.put("agent_run_id", agent_run_id);
            values.put("event_id", event_id);
            values.put("request_id", request_id);
            values.put("correlation_id", correlation_id);
            values.put("taxonomy_version", taxonomy_version);
            return Map.copyOf(values);
        }

        public String toJson() {
            StringBuilder json = new StringBuilder("{");
            boolean first = true;
            for (Map.Entry<String, String> entry : toMap().entrySet()) {
                if (!first) {
                    json.append(',');
                }
                json.append('"').append(entry.getKey()).append("\":\"").append(entry.getValue()).append('"');
                first = false;
            }
            return json.append('}').toString();
        }

        /** Reconstruct a bundle from a decoded JSON object represented as a map. */
        public static Bundle fromMap(Map<String, ?> values) {
            if (values == null || values.size() != 11
                    || !values.keySet().equals(SetNames.ALL)) {
                throw new IllegalArgumentException("identifier payload must contain exactly the canonical fields");
            }
            return new Bundle(
                    stringValue(values, "repository_key"),
                    stringValue(values, "repository_version_id"),
                    stringValue(values, "analysis_id"),
                    stringValue(values, "analysis_job_id"),
                    stringValue(values, "classification_run_id"),
                    stringValue(values, "classification_result_id"),
                    stringValue(values, "agent_run_id"),
                    stringValue(values, "event_id"),
                    stringValue(values, "request_id"),
                    stringValue(values, "correlation_id"),
                    stringValue(values, "taxonomy_version"));
        }

        public static Bundle fromJson(String json) {
            if (json == null || !json.startsWith("{") || !json.endsWith("}")) {
                throw new IllegalArgumentException("identifier payload must be a JSON object");
            }
            Map<String, String> values = new LinkedHashMap<>();
            String body = json.substring(1, json.length() - 1);
            if (!body.isEmpty()) {
                for (String member : body.split(",", -1)) {
                    String[] pair = member.split(":", 2);
                    if (pair.length != 2 || pair[0].length() < 2 || pair[1].length() < 2
                            || pair[0].charAt(0) != '"' || pair[0].charAt(pair[0].length() - 1) != '"'
                            || pair[1].charAt(0) != '"' || pair[1].charAt(pair[1].length() - 1) != '"') {
                        throw new IllegalArgumentException("identifier payload must contain JSON strings");
                    }
                    values.put(pair[0].substring(1, pair[0].length() - 1),
                            pair[1].substring(1, pair[1].length() - 1));
                }
            }
            return fromMap(values);
        }
    }

    private static final class SetNames {
        private static final java.util.Set<String> ALL = java.util.Set.of(
                "repository_key", "repository_version_id", "analysis_id", "analysis_job_id",
                "classification_run_id", "classification_result_id", "agent_run_id", "event_id",
                "request_id", "correlation_id", "taxonomy_version");
    }

    public static String validateRepositoryKey(String value) {
        if (value == null || !REPOSITORY_KEY_PATTERN.matcher(value).matches()) {
            throw new IllegalArgumentException("repository_key must be provider/owner/repository in lowercase");
        }
        return value;
    }

    public static String validateTaxonomyVersion(String value) {
        if (value == null || !TAXONOMY_VERSION_PATTERN.matcher(value).matches()) {
            throw new IllegalArgumentException("taxonomy_version must use vMAJOR.MINOR.PATCH");
        }
        return value;
    }

    public static String validateUuid(String fieldName, String value) {
        if (value == null || !UUID_PATTERN.matcher(value).matches()) {
            throw new IllegalArgumentException(fieldName + " must be a canonical lowercase UUID");
        }
        try {
            UUID.fromString(value);
        } catch (IllegalArgumentException exception) {
            throw new IllegalArgumentException(fieldName + " must be a valid UUID", exception);
        }
        return value;
    }

    public static String deterministicAnalysisKey(String repositoryKey, String commitSha,
                                                  String deterministicPipelineVersion) {
        validateRepositoryKey(repositoryKey);
        validateCommitSha(commitSha);
        validateText("deterministic_pipeline_version", deterministicPipelineVersion);
        return derive("deterministic_analysis", Map.of(
                "commit_sha", commitSha,
                "deterministic_pipeline_version", deterministicPipelineVersion,
                "repository_key", repositoryKey));
    }

    public static String classificationContractId(String classificationPipelineVersion,
                                                  String classificationModelProfile,
                              String promptVersion,
                                                  String taxonomyVersion) {
        validateText("classification_pipeline_version", classificationPipelineVersion);
        validateText("classification_model_profile", classificationModelProfile);
        validateText("prompt_version", promptVersion);
        validateTaxonomyVersion(taxonomyVersion);
        return derive("classification_contract", Map.of(
                "classification_model_profile", classificationModelProfile,
                "classification_pipeline_version", classificationPipelineVersion,
            "prompt_version", promptVersion,
                "taxonomy_version", taxonomyVersion));
    }

    public static String classificationKey(String analysisId, String contractId) {
        validateUuid("analysis_id", analysisId);
        validateText("classification_contract_id", contractId);
        return derive("classification", Map.of(
                "analysis_id", analysisId,
                "classification_contract_id", contractId));
    }

    public static String assembledAnalysisKey(String deterministicKey, String classificationKey) {
        validateText("deterministic_analysis_key", deterministicKey);
        validateText("classification_key", classificationKey);
        return derive("assembled_analysis", Map.of(
                "classification_key", classificationKey,
                "deterministic_analysis_key", deterministicKey));
    }

    public static String agentReviewSpecKey(String repositoryKey, String commitSha, String reviewGoal,
                                            String agentVersion, String promptVersion, String modelProfile) {
        validateRepositoryKey(repositoryKey);
        validateCommitSha(commitSha);
        validateText("review_goal", reviewGoal);
        validateText("agent_version", agentVersion);
        validateText("prompt_version", promptVersion);
        validateText("model_profile", modelProfile);
        return derive("agent_review_spec", Map.of(
                "agent_version", agentVersion,
                "commit_sha", commitSha,
                "model_profile", modelProfile,
                "prompt_version", promptVersion,
                "repository_key", repositoryKey,
                "review_goal", reviewGoal));
    }

    public static String agentContextFingerprint(String classificationResultId, String retrievalIndexVersion,
                                                 String taxonomyVersion, String graphSnapshotVersion) {
        validateUuid("classification_result_id", classificationResultId);
        validateText("retrieval_index_version", retrievalIndexVersion);
        validateTaxonomyVersion(taxonomyVersion);
        validateText("graph_snapshot_version", graphSnapshotVersion);
        return derive("agent_context", Map.of(
                "classification_result_id", classificationResultId,
                "graph_snapshot_version", graphSnapshotVersion,
                "retrieval_index_version", retrievalIndexVersion,
                "taxonomy_version", taxonomyVersion));
    }

    public static String agentResultCacheKey(String reviewSpecKey, String contextFingerprint) {
        validateText("agent_review_spec_key", reviewSpecKey);
        validateText("agent_context_fingerprint", contextFingerprint);
        return derive("agent_result_cache", Map.of(
                "agent_context_fingerprint", contextFingerprint,
                "agent_review_spec_key", reviewSpecKey));
    }

    private static String validateCommitSha(String value) {
        if (value == null || !value.matches("[0-9a-f]{7,64}")) {
            throw new IllegalArgumentException("commit_sha must be 7-64 lowercase hexadecimal characters");
        }
        return value;
    }

    private static String validateText(String fieldName, String value) {
        if (value == null || value.isEmpty()) {
            throw new IllegalArgumentException(fieldName + " must be a non-empty string");
        }
        return value;
    }

    private static String derive(String namespace, Map<String, String> values) {
        Map<String, String> payload = new TreeMap<>(values);
        payload.put("namespace", namespace);
        StringBuilder canonical = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, String> entry : payload.entrySet()) {
            if (!first) {
                canonical.append(',');
            }
            canonical.append(jsonString(entry.getKey())).append(':').append(jsonString(entry.getValue()));
            first = false;
        }
        canonical.append('}');
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                    .digest(canonical.toString().getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder();
            for (byte value : digest) {
                hex.append(String.format("%02x", value));
            }
            return "sha256:" + hex;
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is unavailable", exception);
        }
    }

    private static String jsonString(String value) {
        return "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
    }

    private static String uuid() {
        return UUID.randomUUID().toString();
    }

    private static String stringValue(Map<String, ?> values, String name) {
        Object value = values.get(name);
        if (!(value instanceof String)) {
            throw new IllegalArgumentException(name + " must be a JSON string");
        }
        return (String) value;
    }
}
